lib/Dialect/Seq/Transforms/LowerSeqToSV.cpp

```diff
@@ -136,114 +138,136 @@ void FirRegLower::lower() {
   for (auto reg : llvm::make_early_inc_range(regs))
     toInit.push_back(lower(reg));
 
+  // Compute total width of random space.  Place non-chisel registers at the end
+  // of the space.  The Random space is unique to the initial block, due to
+  // verilog thread rules, so we can drop trailing random calls if they are
+  // unused.
+  uint64_t maxBit = 0;
+  for (auto reg : toInit)
+    if (reg.randStart >= 0)
+      maxBit = std::max(maxBit, (uint64_t)reg.randStart + reg.width);
+  for (auto &reg : toInit)
+    if (reg.randStart == -1) {
+      reg.randStart = maxBit;
+      maxBit += reg.width;
+    }
+
   // Create an initial block at the end of the module where random
   // initialisation will be inserted.  Create two builders into the two
   // `ifdef` ops where the registers will be placed.
   //
-  // `ifdef SYNTHESIS
+  // `ifndef SYNTHESIS
   //   `ifdef RANDOMIZE_REG_INIT
   //      ... regBuilder ...
   //   `endif
   //   initial
   //     `INIT_RANDOM_PROLOG_
   //     ... initBuilder ..
   // `endif
-  if (!toInit.empty()) {
-    auto loc = module.getLoc();
-    MLIRContext *context = module.getContext();
-    auto randInitRef = sv::MacroIdentAttr::get(context, "RANDOMIZE_REG_INIT");
-
-    auto builder =
-        ImplicitLocOpBuilder::atBlockTerminator(loc, module.getBodyBlock());
-    builder.create<sv::IfDefOp>(
-        "SYNTHESIS", [] {},
-        [&] {
-          builder.create<sv::OrderedOutputOp>([&] {
-            auto ifInitOp = builder.create<sv::IfDefOp>(randInitRef);
-            auto regBuilder =
-                ImplicitLocOpBuilder::atBlockEnd(loc, ifInitOp.getThenBlock());
-
-            builder.create<sv::IfDefOp>("FIRRTL_BEFORE_INITIAL", [&] {
-              builder.create<sv::VerbatimOp>("`FIRRTL_BEFORE_INITIAL");
+  if (toInit.empty())
+    return;
+
+  auto loc = module.getLoc();
+  MLIRContext *context = module.getContext();
+  auto randInitRef = sv::MacroIdentAttr::get(context, "RANDOMIZE_REG_INIT");
+
+  auto builder =
+      ImplicitLocOpBuilder::atBlockTerminator(loc, module.getBodyBlock());
+  builder.create<sv::IfDefOp>(
+      "SYNTHESIS", [] {},
+      [&] {
+        builder.create<sv::OrderedOutputOp>([&] {
+          builder.create<sv::IfDefOp>("FIRRTL_BEFORE_INITIAL", [&] {
+            builder.create<sv::VerbatimOp>("`FIRRTL_BEFORE_INITIAL");
+          });
+
+          builder.create<sv::InitialOp>([&] {
+            builder.create<sv::IfDefProceduralOp>("INIT_RANDOM_PROLOG_", [&] {
+              builder.create<sv::VerbatimOp>("`INIT_RANDOM_PROLOG_");
             });
+            llvm::MapVector<Value, SmallVector<RegLowerInfo>> resets;
+            builder.create<sv::IfDefProceduralOp>(randInitRef, [&] {
+              // Create randomization vector
+              SmallVector<Value> randValues;
+              for (uint64_t x = 0; x < (maxBit + 31) / 32; ++x) {
+                auto lhs =
+                    builder.create<sv::LogicOp>(loc, builder.getIntegerType(32),
```

Comment:
Should these be guarded by `ifdef RANDOMIZE_REG_INIT`-- they seem unused otherwise?

EDIT: Oh, they /are/, in the IR anyway.  In the final output they aren't, hmm :thinking: .

Reply:
Yeah, currently Prepare hoists reg and logic without preserving ifdefs.